### ISSUE TITLE
Question on reconnect fix

### ISSUE THREAD

**Issue by ocso** [kongo09/philips-airpurifier-coap#17]
Hi @kongo09,

do you plan to merge "timer" branch back to the master? I have been on "[v0.9.3-beta.1](https://github.com/kongo09/philips-airpurifier-coap/releases/tag/v0.9.3-beta.1)" since you published. Before that release my AC2729 continuously disconnected from HA. After your code change no issue occurred, it reconnects correctly even if I turned off for hours, etc. Values are updated correctly.

**Comment by kongo09** [comment]
That's great to hear. I wasn't entirely sure as I had some problems myself. I'd love to merge it eventually and I'll take another look.

**Comment by ocso** [comment]
Thank you, it seems to be stable for me.

**Comment by GentleGiantSoftware** [comment]
I just stated to use this integration a few days ago and very pleased to see all the work done on this. My AC3829/10 (which is always on) stops being updated in HA after a few hours. If I see the sensor readings are no longer updated I go to the integrations page and do a reload. Sometime this works, but often I only get a single new reading so seems the connection is there, but then again no more updates arrive. I know the connection is unstable, that is clearly documented, but also see that when it works information updates are coming in very quickly, any possibilities to do a manual poll say once a minute or even once every 5 minutes, no need to update the airpurifier information every second?

**Comment by kongo09** [comment]
Unfortunately, to my knowledge, that doesn't work. The device uses the CoAP protocol and while it used to work well in the past, it got very instable after a forced firmware update. I'm not an engineer enough to debug this on protocol level.

Sometimes I have weird effects like readings still working but I cannot send any command. It's really sad.

I tried to contact Philips along several avenues but they consistently don't even reply. It's even more sad as they have launched a cloud API that their own apps but also IFTTT uses, but it is not public.

**Comment by ocso** [comment]
@GentleGiantSoftware, have you tried using "[v0.9.3-beta.1](https://github.com/kongo09/philips-airpurifier-coap/releases/tag/v0.9.3-beta.1)" release? I had the same issue when using other versions.

**Comment by GentleGiantSoftware** [comment]
Hi Osco, in HACS I only see v0.9.6 and later, I was running v0.9.7 just loaded v0.9.7-beta.3 to see if that improves it, or do you suggest to revert specifically to v0.9.3-beta 1?

**Comment by ocso** [comment]
Hi @GentleGiantSoftware,

You need to select "Show beta versions" checkbox to make those releases visible:

![image](https://user-images.githubusercontent.com/10939127/164705526-93c0e4fc-7ee1-4df0-9243-74f91ca343f3.png)

I meant using the specific "v0.9.3-beta.1" release, not the latest.

**Comment by GentleGiantSoftware** [comment]
Hi @ocso, I don't see the v0.9.3-beta.1, I do see other betas.
![image](https://user-images.githubusercontent.com/19820107/164709900-284534c3-a6a8-4bb6-bc75-373e2bbff6f2.png)
Can I just download from github and just overwrite the custom_components folder or will that fail?

**Comment by ocso** [comment]
Yes, you can download and overwrite it, hacs is doing the same.

**Comment by GentleGiantSoftware** [comment]
Hi osco, that seems to be running fine. I now see for instance humidity updates coming in every 2 - 15 seconds (sometimes up to 25 seconds) but at least seems to be running. Will keep you posted, thanks for all the support!

**Comment by kongo09** [comment]
That's great. Please observe this for a few days. I'm still struggeling with the beta and my two devices, so I wasn't sure if I should bring it into the main release, but the more people find it stable, the more confidence this gives. Please also observe your logs for any anomalies or errors.

Oh, and don't try the other betas / alphas. For whatever reason some stuff doesn't work for me locally in my development environment, so sometimes I have to push stuff out as beta and pull it back via HACS to test. The betas are not really running-up to the next release.

**Comment by ocso** [comment]
@kongo09  Just a feedback, I found the following error in a log, only once per day:

```
2022-04-25 07:32:29 ERROR (MainThread) [custom_components.philips_airpurifier_coap.philips] _reconnect error
Traceback (most recent call last):
  File "/config/custom_components/philips_airpurifier_coap/philips.py", line 80, in _reconnect
    self.client = await CoAPClient.create(self._host)
  File "/usr/local/lib/python3.9/site-packages/aioairctrl/coap/client.py", line 38, in create
    await obj._init()
  File "/usr/local/lib/python3.9/site-packages/aioairctrl/coap/client.py", line 33, in _init
    await self._sync()
  File "/usr/local/lib/python3.9/site-packages/aioairctrl/coap/client.py", line 54, in _sync
    response = await self._client_context.request(request).response
asyncio.exceptions.CancelledError
```


```
2022-04-25 07:32:29 ERROR (Recorder) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/aiocoap/tokenmanager.py", line 57, in shutdown
    await self.token_interface.shutdown()
  File "/usr/local/lib/python3.9/site-packages/aiocoap/transports/tcp.py", line 407, in shutdown
    del self._tokenmanager
AttributeError: _tokenmanager

2022-04-25 07:32:29 ERROR (Recorder) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/aiocoap/tokenmanager.py", line 57, in shutdown
    await self.token_interface.shutdown()
  File "/usr/local/lib/python3.9/site-packages/aiocoap/messagemanager.py", line 67, in shutdown
    for messageerror_monitor, cancellable in self._active_exchanges.values():
AttributeError: 'NoneType' object has no attribute 'values'
```


But everything works without issue.

**Comment by kongo09** [comment]
I've done a new beta v0.9.9-timer.2 that is otherwise up-to-date with the rest. Can you check if the error shows with that one as well? Most probably yes...

**Comment by GentleGiantSoftware** [comment]
Hi @ocso, a few days later and v0.9.3-beta 1 is still running fine, sensor readings are being updated and controls work. 
Did not find any log entries with airpurifier or aiocoap in text.
So whatever you did in this version is stable for me!

**Comment by ocso** [comment]
Hi @GentleGiantSoftware,

good to hear that, I have not done anything in code, I just suggested that version as an end-user. It has been done by @kongo09, great work.
I am trying now the new release with this fix: v0.9.9-timer.2

**Comment by kongo09** [comment]
The credit for the timer work goes to @mhetzi https://github.com/kongo09/philips-airpurifier-coap/pull/5

**Comment by ocso** [comment]
@kongo09 I tested v0.9.9-timer.2 branch and caused some known issues like, sensors are updated but device did not accept any command (HA restart solved the problem). I have not found any error or warning in the logs.
So I rollbacked to "[v0.9.3-beta.1](https://github.com/kongo09/philips-airpurifier-coap/releases/tag/v0.9.3-beta.1)"

**Comment by kongo09** [comment]
I know that effect as well, but as far as I understand, it has nothing to do with the timer, but is just one of the messed-up states the device can go into. I believe it keeps broadcasting, but it doesn't accept any direct calls via CoAP protocol.

What the timer might do is allow the sensors to keep going even though the commands don't work anymore, while before everything was gone. But then a HA restart shouldn't be able to fix this.

I've now merged the timer code into the main branch. Hopefully that works for everybody and doesn't create new problems. I really could need some help with all this stuff ...

**Comment by kongo09** [comment]
> any possibilities to do a manual poll

I forgot to answer this, sorry: The manual poll is even more unstable. There is a call in the protocol to get the data. That is actually used during setup to figure out what capabilities the device has and model it is. But sometimes, this just hangs and the device never responds to the call, while it keeps broadcasting its sensor values.

**Comment by kongo09** [comment]
I'll close this now, as timer code is merged and new release v0.10.0 is out
